---
title: "Single-cell and single-nucleus RNA-sequencing from paired normal-adenocarcinoma lung samples provides both common and discordant biological insights"
author: "SÃ©bastien Renaut (sebastien.renaut@criucpq.ulaval.ca)"
date: "`r Sys.time()`"
output:
  html_document:
    number_sections: T
params:
  datapath: 'C:/Users/renseb01/Documents/scRNA/scRNA/data'
  outputpath: '../..' 
---

# Setup  
```{r, setup,echo=FALSE}
knitr::opts_knit$set(root.dir = params$datapath)
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE, cache.lazy=FALSE)
```

```{r, requirements}
library(Seurat) #scRNA analyses
library(sctransform) #data normalization
library(RColorBrewer)
library(dplyr) #data handling
library(patchwork) #data viz
library(ggplot2) #data viz
library(DT) #data viz
library(plotly)

source('../R/azimuth_analysis.R')
source('../R/infercnv.R')
```


# Loading data  
```{r loading,message  = T}  
#seurat.objects =readRDS(file.path(params$datapath,"adeno.24samples_27_07_23cnv.rds"))
filenames =  list.files(params$datapath,pattern = '-C[ST]T-',full.names = F)

metadata = read.csv(file.path(params$datapath,"adeno.24samples_metadata.csv"),row.names = 1)
metadata = relabeler(metadata)
```


# 'SingleCells','SingleNuclei','Depleted' specific plots NORMAL
```{r integration,  include= F}
if(file.exists(file.path(params$datapath,'combined_permethod_pertypeN24.rds')) == F){
combined_permethod = list()
for(i in 1:2){
  if(i == 1) samples = c(1:length(filenames))[regexpr('CST-...._all_',filenames)>0]
  if(i == 2) samples = c(1:length(filenames))[regexpr('CST-...._SN_',filenames)>0]
 
  combined_permethod[[i]] = merge(seurat.objects[[samples[1]]],seurat.objects[samples[-1]],merge.dr = 'proj.umap')
 
  print(paste0('Done ',i,' --- Time is: ', Sys.time()))
}
saveRDS(combined_permethod,file.path(params$datapath,"combined_permethod_pertypeN24.rds"))
} else combined_permethod = readRDS(file.path(params$datapath,'combined_permethod_pertypeN24.rds'))

combined_permethod = relabeler(combined_permethod)

```


# 'SingleCells','SingleNuclei' NORMAL subset epit & Immune
```{r per method, cell cell type normal tissue}
if(file.exists(file.path(params$datapath,'combined_permethod_percelltypeN24.rds')) == F){
type = c('Normal')
cell_type = c('Epithelial','Immune')
combined_permethod_percelltype = list(list(),list())
 for(c in 1:2)
 {
     for(i in 1:2)
    {
      combined_permethod_percelltype[[c]][[i]] = subset(combined_permethod[[i]], subset = predicted.ann_level_1 == cell_type[c])

  category = c('predicted.ann_finest_level','predicted.ann_level_1','predicted.ann_level_2','predicted.ann_level_3','predicted.ann_level_4','predicted.ann_level_5')     
  
    for(cat in seq_along(category)){
      # for annotation's sake, label cells with freq < 1% as 'others' (per sample)
      fine_annotations = combined_permethod_percelltype[[c]][[i]]@meta.data[,colnames(combined_permethod_percelltype[[c]][[i]]@meta.data) %in% category[cat]]
      fine_annotations_summary = table(fine_annotations)
      others = names(fine_annotations_summary)[fine_annotations_summary/sum(fine_annotations_summary)<0.01]
      fine_annotations[fine_annotations %in% others] = 'others'
     
      #fixed the long names 
      fine_annotations = ifelse(nchar(fine_annotations)>12,sub(" ([^ ]*)$", "\n\\1",fine_annotations),fine_annotations)
      
      combined_permethod_percelltype[[c]][[i]]@meta.data[,colnames(combined_permethod_percelltype[[c]][[i]]@meta.data) %in% category[cat]] = fine_annotations
     }
  
}}

saveRDS(combined_permethod_percelltype,file.path(params$datapath,"combined_permethod_percelltypeN24.rds"))
} else combined_permethod_percelltype = readRDS(file.path(params$datapath,'combined_permethod_percelltypeN24.rds'))


combined_permethod_percelltype = relabeler(combined_permethod_percelltype)

```

# Fisher Exact Tests
```{r FETs}
#Fisher Exact Test unknowns epithelials  (level 3)
cell_imm =  table(combined_permethod_percelltype[[2]][[1]]@meta.data$predicted.ann_level_3)
nuc_imm  =  table(combined_permethod_percelltype[[2]][[2]]@meta.data$predicted.ann_level_3)

fisher.test(rbind(c(cell_imm[names(cell_imm) == 'others'],sum(cell_imm[names(cell_imm) != 'others'])),c(nuc_imm[names(nuc_imm) == 'Unclassified'],sum(nuc_imm[names(nuc_imm) != 'Unclassified']))))

#FET  unknowns in general (level 1)
cell =  table(combined_permethod[[1]]@meta.data$predicted.ann_level_1)
nuc =  table(combined_permethod[[2]]@meta.data$predicted.ann_level_1)
fisher.test(rbind(c(cell[names(cell) == 'Unclassified'],sum(cell[names(cell) != 'Unclassified'])),c(nuc[names(nuc) == 'Unclassified'],sum(nuc[names(nuc) != 'Unclassified']))))


#FET epi similar cell types:
cell_epi =  metadata[metadata$type=='Normal',] 
cell_epi =  cell_epi[cell_epi$method=='Cell',]
cell_epi =  cell_epi[cell_epi$predicted.ann_level_1=='Epithelial',]
cell_epi = table(cell_epi$predicted.ann_level_3)


nuc_epi =  metadata[metadata$type=='Normal',] 
nuc_epi =  nuc_epi[nuc_epi$method=='Nucleus',]
nuc_epi =  nuc_epi[nuc_epi$predicted.ann_level_1=='Epithelial',]
nuc_epi = table(nuc_epi$predicted.ann_level_3)


fisher.test(rbind(c(cell_epi[names(cell_epi) == 'Unclassified'],sum(cell_epi[names(cell_epi) != 'Unclassified'])),c(nuc_epi[names(nuc_epi) == 'Unclassified'],sum(nuc_epi[names(nuc_epi) != 'Unclassified']))))
```


# 'SingleCells','SingleNuclei' NORMAL subset epit & ImmunePLOTS
```{r UMAPplots}
###plots
barplots = list()
#color_vectors = list()
cell_type = c('Epithelial','Immune')
for(c in seq_along(cell_type)){

  all_types = c(combined_permethod_percelltype[[c]][[1]]@meta.data$predicted.ann_level_3,
  combined_permethod_percelltype[[c]][[2]]@meta.data$predicted.ann_level_3)
  all_types = names(table(all_types))
  
  
  all_colors = c(brewer.pal(n =12, name = "Paired")[-8],'gray','black','darkred')[1:length(all_types)]
  all_colors[length(all_colors)] = '#FF7F00'#ensure that orange [8] is last for unknown...
  all_colors_type = data.frame(all_types,all_colors)
  
for(i in 1:2){
#cell types plots  FINE LEVELS
  data_summarised = combined_permethod_percelltype[[c]][[i]]@meta.data %>% 
      group_by(predicted.ann_level_3,patient) %>%
      summarise(n = length(predicted.ann_level_1),.groups='keep')

  empty_catego = all_colors_type$all_types
  data_summarised = data_summarised %>% bind_rows(data.frame(predicted.ann_level_3= empty_catego,patient = rep('P1000',length(empty_catego)),n =rep(0,length(empty_catego)) ))#}
  color_vector = all_colors_type$all_colors[all_colors_type$all_types %in% unique(data_summarised$predicted.ann_level_3)]

barplots[[(i*2)+c-2]] = ggplot(data=data_summarised, aes(x='', y=n,fill=predicted.ann_level_3)) +
    geom_bar(stat="identity")+scale_fill_manual( ifelse(c==1,'Cell Types (level3) ~ Epithelial','Cell Types (level3) ~ Immune'),values= color_vector) + ylab('Total Number of Cells')+theme_classic() + xlab('')+theme(axis.title = element_text(size = 8),legend.position = ifelse(i ==1,"none",'right'))
}}

```


#cell types general plots 
```{r celltypes plots}
barplots_general = list()

for(i in 1:2){
  
  data_summarised = combined_permethod[[i]]@meta.data %>% 
      group_by(predicted.ann_level_1,patient) %>%
      summarise(n = length(predicted.ann_level_1),.groups='keep')

barplots_general[[i]] = ggplot(data=data_summarised, aes(x='', y=n,fill=predicted.ann_level_1)) +
    geom_bar(stat="identity")+scale_fill_manual('Cell Types (level1)',values=c(RColorBrewer::brewer.pal(n = 5, name = "Set1")))+theme_classic()+theme(axis.title = element_text(size = 8),legend.position = ifelse(i ==1,"none",'right')) +ylab('Total Number of cells') + xlab('') 
}

```

```{r stats}
table(metadata$method[metadata$predicted.ann_level_1=='Epithelial' & metadata$type == 'Normal'])
```

# plots
```{r plots}
plot_UMAPS = list()

###plots
  plot_SC=DimPlot(combined_permethod[[1]], reduction = "proj.umap", group.by = "predicted.ann_level_1",cols = brewer.pal(n = 5, name = "Set1"),pt.size=0.2) + ggtitle('Cell') + guides(shape='none') +
    theme(legend.position = "none",axis.title = element_text(size = 8),axis.text =element_text(size = 8),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())
 
  plot_SC$labels$colour = 'Cell types (level1)' 
   
  plot_SN=DimPlot(combined_permethod[[2]], reduction = "proj.umap", group.by = "predicted.ann_level_1",cols = brewer.pal(n = 5, name = "Set1"),pt.size=0.2) + ggtitle('Nucleus') + guides(shape='none') +theme(legend.position = "none",axis.title = element_text(size = 8),axis.text =element_text(size = 8),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())

    plot_SN$labels$colour = 'Cell types (level1)'  

###plots
for(c in seq_along(cell_type)){

  all_types = c(combined_permethod_percelltype[[c]][[1]]@meta.data$predicted.ann_level_3,
  combined_permethod_percelltype[[c]][[2]]@meta.data$predicted.ann_level_3)
  all_types = names(table(all_types))
  
  all_colors = c(brewer.pal(n =12, name = "Paired")[-8],'gray','black','darkred')[1:length(all_types)]
  all_colors[length(all_colors)] = '#FF7F00'#ensure that orange [8] is last for unknown...
  all_colors_type = data.frame(all_types,all_colors)
  
for(i in 1:2){
#cell types plots  FINE LEVELS
  data_summarised = combined_permethod_percelltype[[c]][[i]]@meta.data %>% 
      group_by(predicted.ann_level_3,patient) %>%
      summarise(n = length(predicted.ann_level_1),.groups='keep')
  
  color_vector = all_colors_type$all_colors[all_colors_type$all_types %in% unique(data_summarised$predicted.ann_level_3)]
 
  title = paste0(ifelse(i==1,'Cell','Nucleus'),ifelse(c==1,' ~ Epithelial',' ~ Immune'))
  #
  plot_UMAPS[[(i*2)+c-2]] = DimPlot(combined_permethod_percelltype[[c]][[i]], reduction = "proj.umap", cols = color_vector,group.by = "predicted.ann_level_3",pt.size=0.2) + ggtitle(title) + guides(shape='none') + theme(legend.text=element_text(size=7),legend.position='none',axis.title = element_text(size = 8),axis.text =element_text(size = 8),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())
}}
 
#save plots  
  pdf(file.path(params$datapath,'../results/Figures_Oct2023/Figure3_UMAPs_CTA_normal24.pdf'),width = 12,height = 10.5) 
 (plot_SC + barplots_general[[1]] + plot_SN + barplots_general[[2]] + plot_layout(widths = c(10,1,10,1))) / 
 (plot_UMAPS[[2]] + barplots[[2]] + plot_UMAPS[[4]] + (barplots[[4]] + guides(fill=guide_legend(ncol=2))) + plot_layout(widths = c(10,1,10,1)))  / (plot_UMAPS[[1]] + barplots[[1]] + plot_UMAPS[[3]] + (barplots[[3]] + guides(fill=guide_legend(ncol=2))) +plot_layout(widths = c(10,1,10,1))) + plot_annotation(tag_levels = list(c('A','','B','','C','','D','','E','','F','')))
  dev.off()
```

# plots
```{r barplot inspired by Lambrechts et al 2018}
sc_sn_plots = list()

for(i in 1:2){
metadata_simplified = combined_permethod[[i]]@meta.data

metadata_simplified$patient[metadata_simplified$patient == 4] = 1
metadata_simplified$patient[metadata_simplified$patient == 6] = 2
metadata_simplified$patient[metadata_simplified$patient == 8] = 3
metadata_simplified$patient[metadata_simplified$patient == 10] = 4

#none == unknown
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='None'] = 'Unclassified'

#filter rare celltypes
celltypes = as.matrix(table(metadata_simplified$predicted.ann_level_3,metadata_simplified$predicted.ann_level_1))

flush = 0
for(c in 1:ncol(celltypes))
{
  rare = rownames(celltypes)[celltypes[,c]<20]
  #filter 
  if(c != ncol(celltypes)) {temp = c(1:nrow(metadata_simplified))[(metadata_simplified$predicted.ann_level_3 %in% rare) & metadata_simplified$predicted.ann_level_1 == colnames(celltypes)[c]]}
  if(c == ncol(celltypes)) {temp = c(1:nrow(metadata_simplified))[(metadata_simplified$predicted.ann_level_3 == 'Unclassified') & metadata_simplified$predicted.ann_level_1 == colnames(celltypes)[c]]}
  flush = c(temp,flush)
}

metadata_simplified = metadata_simplified[-flush,]

#shorten names
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Lymphatic EC differentiating'] = 'Lymph. EC diff.'
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Lymphatic EC mature'] = 'Lymph. EC matu.'
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Multiciliated lineage'] = 'Multicil. lineage' 
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Innate lymphoid cell NK'] = 'lymphoid cell NK' 

#summarise
data_summarised = metadata_simplified %>% 
  group_by(predicted.ann_level_3,patient,predicted.ann_level_1) %>%
  summarise(n = length(predicted.ann_level_3),sum = sum(nFeature_RNA),cnv_sum = sum(cnv_score),.groups='keep')

data_summarised$n_fraction = 0
data_summarised$nFeature_RNA_mean = 0
data_summarised$cnv_mean = 0
unique_patients = unique(data_summarised$patient)

#keep best labels
for(u in unique(data_summarised$predicted.ann_level_3)){  
  temp = data_summarised[data_summarised$predicted.ann_level_3==u,] 
 
  for(p in seq_along(unique_patients)){
    #this is necessary for the special cases of unknowns which happen in different cell types...
    for(lev1 in c('Immune','Epithelial','Endothelial','Stroma','Unclassified')){ 
      temp$n_fraction[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1] = sum(temp$n[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1]) / sum(temp$n[temp$predicted.ann_level_1==lev1])

      temp$nFeature_RNA_mean[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1] = sum(temp$sum[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])/sum(temp$n[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])
 
      temp$cnv_mean[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1] = sum(temp$cnv_sum[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])/sum(temp$n[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])
    }
  }
  data_summarised[data_summarised$predicted.ann_level_3 ==u,] = temp
}

data_summarised$predicted.ann_level_1 = factor(data_summarised$predicted.ann_level_1,level = c('Immune','Epithelial','Endothelial','Stroma','Unclassified'))

data_summarised$patient = factor(data_summarised$patient,level = c(1,2,3,4))

metadata_simplified$predicted.ann_level_1 = factor(metadata_simplified$predicted.ann_level_1,level = c('Immune','Epithelial','Endothelial','Stroma','Unclassified'))

#plots
fraction_per_patient = ggplot(data_summarised, aes(n_fraction,predicted.ann_level_3,fill = patient)) + 
  geom_col(position = position_fill(reverse = TRUE)) +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
  scale_fill_discrete(type= c(RColorBrewer::brewer.pal(n = 4, name = "Accent"))[4:1]) +
  ylab(paste0('Cell types',ifelse(i==1,' (Cell dataset)',' (Nucleus dataset)'))) +
  xlab('Fraction of cells') +
  theme_bw() + theme(legend.position = ifelse(i==1,'none','bottom'),axis.title.x= list(element_blank(),element_text())[[i]])

number_of_cells = ggplot(data_summarised, aes(n,predicted.ann_level_3 ,fill = patient)) +
    geom_col(position = position_stack(reverse = TRUE)) +
    scale_x_continuous(limits = c(0,ifelse(i==1,12601,6300))) +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
  scale_fill_discrete(type= c(RColorBrewer::brewer.pal(n = 4, name = "Accent")[4:1])) +
  ylab('clusters') +
  xlab('Number of cells') +
   theme_bw() + theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),
  strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])


#data_summarised_UMI = data_summarised[data_summarised$patient=='P4',]
number_of_UMI = ggplot(metadata_simplified, aes(nCount_RNA,predicted.ann_level_3)) + geom_boxplot(fill = "deeppink") + scale_x_continuous(trans= 'log10',breaks = c(500,1500,5000,15000,50000), label = c(500,1500,5000,15000,50000)) +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
  ylab('clusters') +
  xlab('Number of UMIs per cell\n(log scale)') +
   theme_bw() + theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),
  strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])

number_of_genes = ggplot(metadata_simplified, aes(nFeature_RNA,predicted.ann_level_3)) + geom_boxplot(fill = "deeppink") +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
  ylab('clusters') +
  xlab('Number of genes per cell') +
   theme_bw() + theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),
  strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])


#CNV 
cnvMean = ggplot(metadata_simplified, aes(cnv_score,predicted.ann_level_3)) + geom_boxplot(fill = "aquamarine") + scale_x_continuous(trans= 'log10',breaks = c(0.0001,0.001,0.01), label = c(0.0001,0.001,0.01),limits = c(0.00004,0.02)) +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
    ylab('clusters') +
    xlab('CNV score (log scale)') +
    theme_bw() + theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),
                       strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])
  sc_sn_plots[[i]] = fraction_per_patient|number_of_cells|number_of_genes
}

#save plots
pdf(file.path(params$datapath,'../results/Figures_Oct2023/Figure4_barplots_lambrechtsN24.pdf'),width = 11,height = 12) 
(sc_sn_plots[[1]]/sc_sn_plots[[2]] ) + plot_annotation(tag_levels = 'A')
dev.off()
```


```{r done}
print('done')
```