---
title: "Single-cell and single-nucleus RNA-sequencing from paired normal-adenocarcinoma lung samples provides both common and discordant biological insights"
author: "SÃ©bastien Renaut (sebastien.renaut@criucpq.ulaval.ca)"
date: "`r Sys.time()`"
output:
  html_document:
    number_sections: T
params:
  datapath: 'C:/Users/renseb01/Documents/scRNA/scRNA/data'
  outputpath: '../..' 
---

# Setup  
```{r, setup,echo=FALSE}
knitr::opts_knit$set(root.dir = params$datapath)
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE, cache.lazy=FALSE)
```


```{r, requirements}
library(readxl)
library(Seurat) #scRNA analyses
library(RColorBrewer)
library(dplyr) #data handling
library(patchwork) #data viz
library(ggplot2) #data viz
library(DT) #data viz
library(plotly)
library(tidyr)
library(factoextra)
library(FactoMineR)
 library(ggrepel)

source('../R/azimuth_analysis.R')
```

# Loading data  
```{r loading,message  = T}  
seurat.objects =readRDS(file.path(params$datapath,"adeno.24samples_27_07_23cnv.rds"))
seurat.objects = relabeler(seurat.objects)
filenames =  list.files(params$datapath,pattern = '-C[ST]T-',full.names = F)

metadata = read.csv(file.path(params$datapath,"adeno.24samples_metadata.csv"),row.names = 1)
metadata_raw = read.csv(file.path(params$datapath,"adenoRAW.24samples_metadata.csv"),row.names = 1)
metadata = relabeler(metadata)
metadata$unique = paste(metadata$method,metadata$type,metadata$patient,sep = '___')

nb_cells_per_sample = as.data.frame(table(metadata$unique),stringsAsFactors = F)
```


# 'SingleCells','SingleNuclei','Depleted' specific plots NORMAL
```{r integration,  include= F}
if(file.exists(file.path(params$datapath,'combined_depletion_pertissue.rds')) == F){
combined_permethod = list()
for(i in 1:2){
  if(i == 1) samples = c(1:length(filenames))[regexpr('CST-...._depletion_',filenames)>0]
  if(i == 2) samples = c(1:length(filenames))[regexpr('CTT-...._depletion_',filenames)>0]
 
  combined_permethod[[i]] = merge(seurat.objects[[samples[1]]],seurat.objects[samples[-1]])
}
saveRDS(combined_permethod,file.path(params$datapath,"combined_depletion_pertissue.rds"))
} else combined_permethod = readRDS(file.path(params$datapath,'combined_depletion_pertissue.rds'))

#The relabeler is just there to change all UNKNOWNS -> unclassified.
combined_permethod = relabeler(combined_permethod)
```


#cell types & UMAPs
```{r celltypes plots}
barplots_general = list()
for(i in 1:2){
  
  data_summarised = combined_permethod[[i]]@meta.data %>% 
      group_by(predicted.ann_level_1,patient) %>%
      summarise(n = length(predicted.ann_level_1),.groups='keep')

barplots_general[[i]] = ggplot(data=data_summarised, aes(x='', y=n,fill=predicted.ann_level_1)) +
    geom_bar(stat="identity")+scale_fill_manual('Cell Types (level 1)',values=c(RColorBrewer::brewer.pal(n = 5, name = "Set1")))+theme_classic()+theme(axis.title = element_text(size = 8),legend.position = ifelse(i ==1,"none",'right')) +ylab('Total Number of cells') + xlab('') 
}

plot_UMAPS = list()
###plots
  plot_depletion_N=DimPlot(combined_permethod[[1]], reduction = "proj.umap", group.by = "predicted.ann_level_1",cols = brewer.pal(n = 5, name = "Set1"),pt.size=0.1) + ggtitle('Immune depleted cell ~ Normal') + guides(shape='none') +theme(legend.position = "none",axis.title = element_text(size = 8),axis.text =element_text(size = 8),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())
  
   plot_depletion_N$labels$colour = 'Cell types ~ level1'  
   
   
  plot_depletion_T=DimPlot(combined_permethod[[2]], reduction = "proj.umap", group.by = "predicted.ann_level_1",cols = brewer.pal(n = 5, name = "Set1"),pt.size=0.1) + ggtitle('Immune depleted cell ~ Tumor') + guides(shape='none') +theme(legend.position = "none",axis.title = element_text(size = 8),axis.text =element_text(size = 8),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())
  plot_depletion_T$labels$colour = 'Cell types ~ level1'  

```

#percentage mitochondrial as a function of method and tissue type
```{r percentmt,message  = T,echo=T} 
data_summarised = metadata_raw %>% group_by(method,type,samplename) %>% summarise(percent.mt = mean(percent.mt))
metadata_raw$method[metadata_raw$method=='immune depleted cells'] = 'Immune depleted\ncells'
p_mt_method_type = ggplot(metadata_raw, aes(x=method, y=percent.mt, fill=method)) +
  geom_boxplot(position="dodge",outlier.shape = NA) +
  ylab("Percentage of mito-\nchondrial UMIs")  +
  facet_wrap(~type)+ylim(0,40)+
  theme_bw()+
  xlab('') +
  theme(axis.text.x = element_blank(),legend.position = "right") 

#anova for stats
anova(lm(percent.mt~method*type+samplename, data = data_summarised))


```


# STRESS ANALYSIS
```{r loadRDS,message  = T,echo=T}  
#got this file from OFlanagan_et_al2019. 
stress_references = read_xlsx(file.path(params$datapath,"512genes_OFlanagan_et_al2019.xlsx"))

#prepare output df
stress_genes = data.frame(gene = rownames(seurat.objects[[i]]@assays$RNA)[rownames(seurat.objects[[i]]@assays$RNA) %in% stress_references$gene_symbol])

for(i in 1:24){
  #stress_genesXcell is a matrix of stress genes X cells
  stress_genesXcell  = seurat.objects[[i]]@assays$RNA$counts[rownames(seurat.objects[[i]]@assays$RNA) %in% stress_references$gene_symbol,]
  temp2 = NULL

  for(j in 1:nrow(stress_genesXcell)){
    #
    temp2 = c(temp2,length(stress_genesXcell[j,stress_genesXcell[j,]>0])/ncol(stress_genesXcell))
    }
  stress_genes = cbind(stress_genes,temp2)
}

#
colnames(stress_genes)[-1] = filenames
stress_genes = stress_genes %>%
    pivot_longer(!gene, names_to = "filenames")

#
stress_genes$method  = '0'
stress_genes$type = '0'
stress_genes$patient = '0'

stress_genes$type = ifelse(regexpr('CST',stress_genes$filenames)>0,'Normal','Tumor')
stress_genes$method[regexpr('_SN_',stress_genes$filenames)>0] = 'Nucleus'
stress_genes$method[regexpr('_all_',stress_genes$filenames)>0] = 'Cell'
stress_genes$method[regexpr('_depletion_',stress_genes$filenames)>0] = 'Immune\ndepleted cell'

stress_genes$patient[regexpr('^4',stress_genes$filenames)>0] = 'P4'
stress_genes$patient[regexpr('^6',stress_genes$filenames)>0] = 'P6'
stress_genes$patient[regexpr('^8',stress_genes$filenames)>0] = 'P8'
stress_genes$patient[regexpr('^10',stress_genes$filenames)>0] = 'P10'
stress_genes$site = 'LeftUpperLobe' 
stress_genes$site[stress_genes$patient=='P8'] = 'RightUpperLobe'
stress_genes$site[stress_genes$patient=='P10'] = 'RightUpperLobe'

stress_genes$sqrt = stress_genes$value**0.5

#statistical test
anova(lm(value~method*type+patient, data = stress_genes))

stress_genes %>% 
  group_by(type,method) %>%
  summarise(n = median(value,na.rm =T),.groups='keep') 

stress_genes %>% 
  group_by(type) %>%
  summarise(n = median(value,na.rm =T),.groups='keep') 

stress_genes %>% 
  group_by(method) %>%
  summarise(n = median(value,na.rm =T),.groups='keep') 


stress_genes %>% 
  group_by(patient) %>%
  summarise(n = median(value,na.rm =T),.groups='keep') 

###plots
p_method = ggplot(stress_genes, aes(x=method, y=value, fill=method)) +
  geom_boxplot(position="dodge",outlier.shape = NA) +
  ylab("mean % of cells expressing a stress-related gene")  +
  theme(axis.text.x = element_text(size=7))

p_type = ggplot(stress_genes, aes(x=type, y=value, fill=type)) +
  geom_boxplot(position="dodge",outlier.shape = NA) +
  scale_fill_manual(values = c('yellow','darkblue')) +
  ylab("")

p_patient = ggplot(stress_genes, aes(x=patient, y=value, fill=patient)) +
  geom_boxplot(position="dodge",outlier.shape = NA) +
  ylab('')

p_site = ggplot(stress_genes, aes(x=site, y=value, fill=method)) +
  geom_boxplot(position="dodge",outlier.shape = NA) +
  facet_wrap(~type)+
  ylab("fraction of cells expressing stress-related gene")


p_method_type = ggplot(stress_genes, aes(x=method, y=value, fill=method)) +
  geom_boxplot(position="dodge",outlier.shape = NA) +
  ylab("fraction of cells expressing\na stress-related gene")  +
  facet_wrap(~type)+
  theme_bw() +
  xlab('') +
  theme(axis.text.x = element_blank(),legend.position = "none") 
```  

# Cell types annotations summaries
```{r celltypes,echo = F} 
data_summarised = metadata %>% 
    group_by(predicted.ann_level_1,method,type,patient,unique) %>%
    summarise(n = length(predicted.ann_level_1),.groups='keep')

data_summarised$fraction = 0

for(i in seq_along(nb_cells_per_sample[,1])){
  data_summarised$fraction[data_summarised$unique == nb_cells_per_sample$Var1[i]] =  data_summarised$n[data_summarised$unique == nb_cells_per_sample$Var1[i]] / nb_cells_per_sample$Freq[i]*100
}

data_summarised$methodXtype = paste(data_summarised$type,data_summarised$method,sep = ' --- ')

data_summarised$patient = factor(data_summarised$patient,level = c(4,6,8,10))

summary_bars = ggplot(data=data_summarised, aes(x=patient, y=fraction, fill=predicted.ann_level_1)) +
  geom_bar(stat="identity") + facet_wrap(~interaction(method,type,sep = " ~ "),nrow = 2, ncol = 3, scales = "free_y") + 
  scale_fill_manual('Cell Types (level1)',values=c(RColorBrewer::brewer.pal(n = 5, name = "Set1")))+xlab('Patients') + ylab('Percentage of cells')
```

# CNV feature plot
```{r cnv,echo = F} 
combined_permethod[[2]]@meta.data$cnv_score[combined_permethod[[2]]@meta.data$cnv_score>0.015] = 0.015 
cnv_depletion_N= FeaturePlot(combined_permethod[[2]],order = T,reduction = 'proj.umap',features = 'cnv_score',pt.size = 0.2,cols = c('#D3D3D3ff','#ff0000ff')) + ggtitle('Immune depleted cell (Tumor) ~ CNV score') + guides(shape='none') + theme(legend.position = "right",axis.title = element_text(size = 8),axis.text =element_text(size = 8) )

cnv_depletion_N$labels$colour = 'score'  
```


```{r barplot inspired by Lambrechts et al 2018}
sc_sn_plots = list()

#here i is to do the analysis plot for Cell vs. Nuclei.
for(i in 1:2){
metadata_simplified = combined_permethod[[i]]@meta.data
metadata_simplified$patient[metadata_simplified$patient == 4] = 1
metadata_simplified$patient[metadata_simplified$patient == 6] = 2
metadata_simplified$patient[metadata_simplified$patient == 8] = 3
metadata_simplified$patient[metadata_simplified$patient == 10] = 4
#none == unknown
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='None'] = 'Unclassified'

#filter rare celltypes
celltypes = as.matrix(table(metadata_simplified$predicted.ann_level_3,metadata_simplified$predicted.ann_level_1))

flush = 0
for(c in 1:ncol(celltypes))
{
  rare = rownames(celltypes)[celltypes[,c]<20]
  #filter 
  if(c != ncol(celltypes)) {temp = c(1:nrow(metadata_simplified))[(metadata_simplified$predicted.ann_level_3 %in% rare) & metadata_simplified$predicted.ann_level_1 == colnames(celltypes)[c]]}
  if(c == ncol(celltypes)) {temp = c(1:nrow(metadata_simplified))[(metadata_simplified$predicted.ann_level_3 == 'Unclassified') & metadata_simplified$predicted.ann_level_1 == colnames(celltypes)[c]]}
  flush = c(temp,flush)
}

metadata_simplified = metadata_simplified[-flush,]

#shorten names for visualisation
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Lymphatic EC differentiating'] = 'Lymph. EC diff.'
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Lymphatic EC mature'] = 'Lymph. EC matu.'
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Multiciliated lineage'] = 'Multicil. lineage' 
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='Innate lymphoid cell NK'] = 'lymphoid cell NK'
metadata_simplified$predicted.ann_level_3[metadata_simplified$predicted.ann_level_3=='SM activated stress response'] = 'stress response'

#summarise
data_summarised = metadata_simplified %>% 
  group_by(predicted.ann_level_3,patient,predicted.ann_level_1) %>%
  summarise(n = length(predicted.ann_level_3),sum = sum(nFeature_RNA),cnv_sum = sum(cnv_score),.groups='keep')

data_summarised$n_fraction = 0
data_summarised$nFeature_RNA_mean = 0
data_summarised$cnv_mean = 0
unique_patients = unique(data_summarised$patient)

#keep best labels
for(u in unique(data_summarised$predicted.ann_level_3)){  
  temp = data_summarised[data_summarised$predicted.ann_level_3==u,] 
 
  for(p in seq_along(unique_patients)){
    #this is necessary for the special cases of unknowns which happen in different cell types...
    for(lev1 in c('Immune','Epithelial','Endothelial','Stroma','Unclassified')){ 
      temp$n_fraction[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1] = sum(temp$n[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1]) / sum(temp$n[temp$predicted.ann_level_1==lev1])

      temp$nFeature_RNA_mean[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1] = sum(temp$sum[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])/sum(temp$n[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])
 
      temp$cnv_mean[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1] = sum(temp$cnv_sum[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])/sum(temp$n[temp$patient==unique_patients[p] & temp$predicted.ann_level_1 == lev1])
    }
  }
  data_summarised[data_summarised$predicted.ann_level_3 ==u,] = temp
}

data_summarised$predicted.ann_level_1 = factor(data_summarised$predicted.ann_level_1,level = c('Immune','Epithelial','Endothelial','Stroma','Unclassified'))

data_summarised$patient = factor(data_summarised$patient,level = c(1,2,3,4))

metadata_simplified$predicted.ann_level_1 = factor(metadata_simplified$predicted.ann_level_1,level = c('Immune','Epithelial','Endothelial','Stroma','Unclassified'))


#plots
fraction_per_patient = ggplot(data_summarised, aes(n_fraction,predicted.ann_level_3,fill = patient)) + 
  geom_col(position = position_fill(reverse = TRUE)) +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
  xlab('Fraction of cells') +
  scale_fill_discrete(type= c(RColorBrewer::brewer.pal(n = 4, name = "Accent"))[4:1]) +
  theme_bw() + theme(legend.position = ifelse(i==1,'right','none'),axis.title.x= list(element_blank(),element_text())[[i]])

number_of_cells = ggplot(data_summarised, aes(n,predicted.ann_level_3,fill = patient)) +
    geom_col(position = position_stack(reverse = TRUE))  +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
    xlim(c(0,8500)) + 
  ylab(paste0('Cell types',ifelse(i==1,' (Normal)',' (Tumor)')))  +
  xlab('Number of cells') +
  scale_fill_discrete(type= c(RColorBrewer::brewer.pal(n = 4, name = "Accent"))[4:1]) +
  theme_bw() + theme(legend.position = ifelse(i==1,'none','right'),strip.text = element_text(
    size = 8),axis.text.y = element_text(size = 7))

#data_summarised_UMI = data_summarised[data_summarised$patient=='P4',]
number_of_UMI = ggplot(metadata_simplified, aes(nCount_RNA,predicted.ann_level_3)) + geom_boxplot(fill = "deeppink") + 
  facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') + 
  scale_x_continuous(trans = 'log10', breaks = c(500,1500,5000,15000,50000), label = c(500,1500,5000,15000,50000)) +
  xlab('Number of UMIs per cell\n(log scale)') +
  theme_bw() + 
  theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])


#CNV 
number_of_genes = ggplot(metadata_simplified, aes(nFeature_RNA,predicted.ann_level_3)) + geom_boxplot(fill = "deeppink") +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
  ylab('clusters') +
  xlab('Number of genes per cell') +
   theme_bw() + theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),
  strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])

cnvMean = ggplot(metadata_simplified, aes(cnv_score,predicted.ann_level_3)) + geom_boxplot(fill = "aquamarine") + scale_x_continuous(trans= 'log10',breaks = c(0.0001,0.001,0.01), label = c(0.0001,0.001,0.01),limits = c(0.00004,0.02)) +
    facet_grid(predicted.ann_level_1 ~ ., scales = "free", space = "free",switch = 'y') +
    ylab('clusters') +
    xlab('CNV score (log scale)') +
    theme_bw() + theme(legend.position = 'none',strip.background = element_blank(),axis.text.y=element_blank(),
                       strip.text.y = element_blank(),axis.title.y=element_blank(),axis.title.x= list(element_blank(),element_text())[[i]])

 if(i ==1) sc_sn_plots[[i]] = number_of_cells
 if(i ==2) sc_sn_plots[[i]] = number_of_cells
}


#pdf(file.path(params$datapath,'../results/Figures_august2023/Suppl_Fig9_barplots_lambrechtsDepletion.pdf'),width = 14,height = 12) 
(sc_sn_plots[[1]]/sc_sn_plots[[2]] ) + plot_annotation(tag_levels = list(c('A','B','C','D','E','F','G')))
#dev.off()
```

```{r save plots, message= T}
#save plots
  pdf(file.path(params$datapath,'../results/Figures_March2024/Figure10_UMAPs_depletion24.pdf'),width = 12,height = 9) 
 ((plot_depletion_N + barplots_general[[1]] + plot_depletion_T + barplots_general[[2]] + plot_layout(widths = c(10,1,10,1))) / 
 
  (sc_sn_plots[[1]]|sc_sn_plots[[2]] | ((p_method_type / p_mt_method_type)  +     plot_layout(guides = 'collect')        )      ) + plot_layout(widths = c(6,6))) + plot_annotation(tag_levels = list(c('A','','B','','C','D','E','F')))
          #                                                                                                                                 )
  dev.off()
```


# PCA top5% most variable genes
```{r pcas,message  = T}  
combined_sub = merge(seurat.objects[[1]],seurat.objects[c(2:24)])

# top 5percent most variables
combined_agg_DF2  <- AggregateExpression(combined_sub, assays = "SCT", return.seurat = F, group.by = c("type",'method',"patient"))
gexpr = as.data.frame(combined_agg_DF2$SCT)
gexpr = log(gexpr+1,2)

gexpr_sd = apply(gexpr,1,sd)

#cutoffs
cutoff = quantile(gexpr_sd,probs = seq(0,1,by = 0.05))[20]
top5percent_variable = names(gexpr_sd)[gexpr_sd>=cutoff]
seurat.objects.subsetted = list()

#subset only the top 5% most variable
for(i in 1:24){
  seurat.objects.subsetted[[i]] = subset(seurat.objects[[i]],features = top5percent_variable)
}

#combine
combined_raw = merge(seurat.objects.subsetted[[1]],seurat.objects.subsetted[c(2:24)])
combined = subset(combined_raw, subset = predicted.ann_level_1 != 'Unclassified')
combined = subset(combined, subset = type != 'Tumor')

combined_agg_DF = AggregateExpression(combined, assays = "SCT", return.seurat = F, group.by = c('method', "patient"))
combined_agg_DF = data.frame(combined_agg_DF$SCT,check.names = F)


#normalise by the number of cells
number_of_cells = as.data.frame(table(combined@meta.data$method,combined@meta.data$patient))
number_of_cells = number_of_cells[order(number_of_cells$Var1,number_of_cells$Var2),]
combined_agg_DF = (apply(combined_agg_DF, 1, function(x) x/number_of_cells$Freq))

#PCA
pseudobulkRNAseq_pca = PCA(combined_agg_DF,scale = T, graph = F) #I just dont want the None category

#samples DF
data = as.data.frame(pseudobulkRNAseq_pca$ind$coord)
rownames(data) = gsub('_4','_1',rownames(data))
rownames(data) = gsub('_6','_2',rownames(data))
rownames(data) = gsub('_8','_3',rownames(data))
rownames(data) = gsub('_10','_4',rownames(data))
data$Patients = sapply(strsplit(rownames(data), "_",fixed = T), "[",2)
data$method = sapply(strsplit(rownames(data), "_",fixed = T), "[",1)
data$`method X Patients` = rownames(data)


#gene variable DF
plot_depleted_all = ggplot(data,aes(x= Dim.1,y=Dim.2, fill = `method X Patients` )) + 
  geom_hline(yintercept=0,linetype="dashed") + 
  geom_vline(xintercept=0,linetype="dashed") + 
  geom_point(size = 6,color = 'black', shape=21,alpha =0.75)  + 
  ylab('PC2') +
  xlab('PC1') +
  theme_bw() + 
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu")[c(4:1,8:11)],RColorBrewer::brewer.pal(n = 9, name = "Greens")[c(3,5,7,9)])  ) +
  theme(legend.position = 'right') +
  ggtitle('2,209 (top5%) most variable genes')
```



# PCA marker genes
```{r pca marker genes, message= T}
###session
Immune = c('LCP1','CD53','PTPRC','COTL1','CXCR4','GMFG','FCER1G','LAPTM5','SRGN','CD52')
Epithelial = c('KRT7','PIGR','ELF3','CYB5A','KRT8','KRT19','TACSTD2','MUC1','S100A14','CXCL17')
Endothelial = c('PTRF','CLDN5','AQP1','PECAM1','NPDC1','VWF','GNG11','RAMP2','CLEC14A')
Stroma = c('TPM2','DCN','MGP','SPARC','CALD1','LUM','TAGLN','IGFBP7','COL1A2','C1S','supetoto')

features = unique(c(Immune,Epithelial,Endothelial,Stroma))

seurat.objects.subsetted = list()

for(i in 1:24){
  seurat.objects.subsetted[[i]] = subset(seurat.objects[[i]],features = features)
}

#combine
combined = merge(seurat.objects.subsetted[[1]],seurat.objects.subsetted[c(2:24)])
combined = subset(combined, subset = predicted.ann_level_1 != 'Unclassified')
combined = subset(combined, subset = type != 'Tumor')
combined@meta.data$method = gsub('immune','Immune',combined@meta.data$method)

combined_agg_DF = AggregateExpression(combined, assays = "SCT", return.seurat = F, group.by = c('method', "patient","predicted.ann_level_1"))
combined_agg_DF = data.frame(combined_agg_DF$SCT,check.names = F)

# each 'cell' is a donor-condition-celltype pseudobulk profile
combined_agg_DF = combined_agg_DF[,order(colnames(combined_agg_DF))]

#normalise by the number of cells
number_of_cells = as.data.frame(table(combined@meta.data$patient,combined@meta.data$predicted.ann_level_1,combined@meta.data$method))
number_of_cells$type_method_level1 = paste(number_of_cells$Var3,number_of_cells$Var1,number_of_cells$Var2,sep = '_')
number_of_cells = number_of_cells[order(number_of_cells$type_method_level1),]

combined_agg_DF = (apply(combined_agg_DF, 1, function(x) x/number_of_cells$Freq))

#PCA
pseudobulkRNAseq_pca = PCA(combined_agg_DF,scale = T, graph = F) #I just dont want the None category

#samples DF
data = as.data.frame(pseudobulkRNAseq_pca$ind$coord)
rownames(data) = gsub('_4','_1',rownames(data))
rownames(data) = gsub('_6','_2',rownames(data))
rownames(data) = gsub('_8','_3',rownames(data))
rownames(data) = gsub('_10','_4',rownames(data))

data$Patients = sapply(strsplit(rownames(data), "_",fixed = T), "[",2)
#data$tissue = sapply(strsplit(rownames(data), "_",fixed = T), "[",1)
data$method = sapply(strsplit(rownames(data), "_",fixed = T), "[",1)
data$`cell types` = sapply(strsplit(rownames(data), "_",fixed = T), "[",3)

p_tissuemethod = ggplot(data,aes(x= Dim.1,y=Dim.2,shape = method, fill = `cell types`)) +
    geom_hline(yintercept=0,linetype="dashed") + 
    geom_vline(xintercept=0,linetype="dashed") + 
    geom_point(size = 4,alpha =0.75)  + 
    ylab('PC2') +
    xlab('PC1') +
    scale_fill_manual(values = brewer.pal(n = 4, name = "Set1")) +
    scale_color_manual(values = brewer.pal(n = 4, name = "Set1")) +
    scale_shape_manual(values=c(21, 22, 24)) +
    theme_bw() +
    theme(legend.position = 'right') +
    ggtitle('Cell type canonical markers (38 genes)')+
    guides(fill = guide_legend(override.aes = list(shape= 21,fill = brewer.pal(n = 4, name = "Set1"))))
```


# PCA cell types
```{r pca cell types, message= T}
#PC with celltypes
ct = as.data.frame(table(metadata$patient,metadata$type,metadata$method,metadata$predicted.ann_level_3))
ct = ct[ct$Var4!= 'Unclassified',]
ct = ct[ct$Var4!= 'Submucosal Secretory',]
ct = ct[ct$Var2!= 'Tumor',]
ct = ct[ct$Var4!='None', ]

ct$names = paste0(ct$Var3,'_',ct$Var1)
ct = tidyr::pivot_wider(ct,values_from= Freq,names_from = Var4)
ct_normalised = as.data.frame(ct[,-c(1:4)])
rownames(ct_normalised) = ct$names
ct_normalised = t(ct_normalised)
number_of_cells = colSums(ct_normalised)

ct_normalised = (apply(ct_normalised, 1, function(x) x/number_of_cells))

#PCA
pseudobulkRNAseq_pca = PCA(ct_normalised,scale = T, graph = F) #I just dont want the None category
data = as.data.frame(pseudobulkRNAseq_pca$ind$coord)
rownames(data) = gsub('_4','_1',rownames(data))
rownames(data) = gsub('_6','_2',rownames(data))
rownames(data) = gsub('_8','_3',rownames(data))
rownames(data) = gsub('_10','_4',rownames(data))
data$`Method X sample` = rownames(data)

#gene variable DF
gene_variables = as.data.frame(pseudobulkRNAseq_pca$var)[,1:4]
gene_variables$celltypes = rownames(gene_variables)

#
p_celltype  = ggplot() + 
#    geom_segment(data =gene_variables,  aes(x = rep(0,21), y = rep(0,21), xend = coord.Dim.1*max(abs(data$Dim.1)), yend = coord.Dim.2*max(abs(data$Dim.2))), arrow = arrow(length = unit(0.2, "cm")), show.legend = T ) + 
    geom_hline(yintercept=0,linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") +
#  geom_label_repel(data = gene_variables, aes(label = celltypes, x = coord.Dim.1*max(abs(data$Dim.1)), y = coord.Dim.2*max(abs(data$Dim.2))),size = 2) +
  geom_point(data=data,aes(x= Dim.1,y=Dim.2, fill = `Method X sample`), size = 6,color = 'black', shape = 21,alpha =1) +
  ylab('PC2') +
  xlab('PC1') +
  theme_bw() + 
  scale_fill_manual(values = c(brewer.pal(n = 11, name = "RdYlBu")[c(4:1,8:11)],brewer.pal(n = 9, name = "Greens")[c(3,5,7,9)])  ) +
  theme(legend.position = 'right') +
  ggtitle('Cell types (level3)')
#
```


```{r plots}

pdf(file.path(params$datapath,'../results/Figures_March2024/Fig_S10_depletedPCA.pdf'),width = 12,height = 8) 
(plot_depleted_all+ p_celltype)  / ( p_tissuemethod+ plot_spacer()) + plot_annotation(tag_levels = 'A')
dev.off()


```













```


