---
title: "Single-cell and single-nucleus RNA-sequencing of the same paired normal-adenocarcinoma lung samples provides divergent biological information"
author: "SÃ©bastien Renaut (sebastien.renaut@criucpq.ulaval.ca)"
date: "`r Sys.time()`"
output:
  html_document:
    number_sections: T
params:
  datapath: 'C:/Users/renseb01/Documents/scRNA/scRNA/data'
  outputpath: '../..' 
---

# Setup  
```{r, setup,echo=FALSE}
knitr::opts_knit$set(root.dir = params$datapath)
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, cache=FALSE, cache.lazy=FALSE)
```

```{r, requirements}
library(Seurat) #scRNA analyses
library(RColorBrewer)
library(dplyr) #data handling
library(patchwork) #data viz
library(ggplot2) #data viz
library(DT) #data viz
library(plotly)

source('../R/azimuth_analysis.R')
source('../R/infercnv.R')
```


# Loading data  
```{r loading,message  = T, fig.width = 11, fig.height=8,echo=T}  
seurat.objects =readRDS(file.path(params$datapath,"adeno.24samples_27_07_23cnv.rds"))

seurat.objects = relabeler(seurat.objects)

filenames =  list.files(params$datapath,pattern = '-C[ST]T-',full.names = F)

metadata = read.csv(file.path(params$datapath,"adeno.24samples_metadata.csv"),row.names = 1)

metadata = relabeler(metadata)
```


# get data ready
```{r session, message= T}
Immune = c('LCP1','CD53','PTPRC','COTL1','CXCR4','GMFG','FCER1G','LAPTM5','SRGN','CD52')
Epithelial = c('KRT7','PIGR','ELF3','CYB5A','KRT8','KRT19','TACSTD2','MUC1','S100A14','CXCL17')
Endothelial = c('PTRF','CLDN5','AQP1','PECAM1','NPDC1','VWF','GNG11','RAMP2','IGFBP7','CLEC14A')
Stroma = c('TPM2','DCN','MGP','SPARC','CALD1','LUM','TAGLN','IGFBP7','COL1A2','C1S')
features = c('EPCAM', 'CLDN5', 'COL1A2','PTPRC')

random100 = sample(seurat.objects[[1]]@assays$refAssay@data@Dimnames[[1]],100)
  
  
all_features = unique(c(Immune,Epithelial,Endothelial,Stroma,features))#,random100))
seurat.objects.subsetted = list()

for(i in 1:24){
  seurat.objects.subsetted[[i]] = subset(seurat.objects[[i]],features = all_features)
}

#combine
combined = merge(seurat.objects.subsetted[[1]],seurat.objects.subsetted[c(2:24)])

#
combined@meta.data$predicted.ann_level_3_simplified = combined@meta.data$predicted.ann_level_3
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3 %in% c('SM activated stress response','Submucosal Secretory')] = 'others'
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3 == 'None'] = 'Unclassified'
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3 == 'unknown'] = 'Unclassified'

#
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3_simplified == 'Unclassified' & combined@meta.data$predicted.ann_level_1 == 'Epithelial'] = 'Epithelial ~ Unclassified'
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3_simplified == 'Unclassified' & combined@meta.data$predicted.ann_level_1 == 'Endothelial'] = 'Endothelial ~ Unclassified'
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3_simplified == 'Unclassified' & combined@meta.data$predicted.ann_level_1 == 'Immune'] = 'Immune ~ Unclassified'
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3_simplified == 'Unclassified' & combined@meta.data$predicted.ann_level_1 == 'Stroma'] = 'Stroma ~ Unclassified'

#
combined@meta.data$patient[combined@meta.data$patient == 4] = 1
combined@meta.data$patient[combined@meta.data$patient == 6] = 2
combined@meta.data$patient[combined@meta.data$patient == 8] = 3
combined@meta.data$patient[combined@meta.data$patient == 10] = 4

#
combined@meta.data$method[combined@meta.data$method == 'immune depleted cells'] = 'Immune-depleted\ncell'
combined@meta.data$predicted.ann_level_3_simplified[combined@meta.data$predicted.ann_level_3_simplified=='Lymphatic EC differentiating'] = 'Lymphatic EC diff.'
combined@meta.data$predicted.ann_level_3_simplified = factor(combined@meta.data$predicted.ann_level_3_simplified, levels = c("AT1","AT2","B cell lineage","Basal","Dendritic cells","EC arterial","EC capillary","EC venous","Fibroblasts","Innate lymphoid cell NK","Lymphatic EC diff.","Lymphatic EC mature","Macrophages","Mast cells","Monocytes","Multiciliated lineage","Myofibroblasts","others","Rare","Secretory","T cell lineage","Endothelial ~ Unclassified","Epithelial ~ Unclassified","Immune ~ Unclassified","Stroma ~ Unclassified","Unclassified"))
```


# Large UMAP of all cells.
```{r plots, message= T}
col = c(brewer.pal(n = 8, name = "Set1")[6:8],brewer.pal(n = 7, name = "Set3"),'black','darkblue','darkred',brewer.pal(8,name = 'Accent')[8:1],brewer.pal(n = 8, name = "Set1")[1:5])

umap_theme = function(){
  theme(legend.position = "right",axis.title = element_text(size = 16),axis.text = element_text(size = 16),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank()) 
}

dimplots_allcells = DimPlot(combined, reduction = "proj.umap", group.by = "predicted.ann_level_3_simplified",cols = col,pt.size=0.01,raster= F,order = F, label = T, label.size = 8, repel = T) + 
    ggtitle('All cells/nuclei ~ Annotation (level 3)') +
    guides(shape='none',color=guide_legend(override.aes = list(size=10), ncol=1)) + 
    labs(color = "Cell Types (level3)") +
    umap_theme() + theme(plot.title = element_text(size = 42),legend.text = element_text(size=16),legend.title = element_text(size=28), legend.position = 'right')

dimplots_type = DimPlot(combined, reduction = "proj.umap", group.by = "type",pt.size=0.01,raster= F,order = F, label.size = 4) + 
    ggtitle('All cells/nuclei ~ Tissue type') +
    guides(shape='none',color=guide_legend(override.aes = list(size=10), ncol=1)) +
    labs(color = "Tissue type") +
    umap_theme() + theme(plot.title = element_text(size = 42),legend.text = element_text(size=20),legend.title = element_text(size=28), legend.position = 'right')

dimplots_method = DimPlot(combined, reduction = "proj.umap", group.by = "method",pt.size=0.01,raster= F,order = F, label.size = 8) + 
    ggtitle('All cells/nuclei ~ Experimental method') +
    guides(shape='none',color=guide_legend(override.aes = list(size=10), ncol=1)) +
    labs(color = "Method") +
    umap_theme() + theme(plot.title = element_text(size = 42),legend.text = element_text(size=20),legend.title = element_text(size=28), legend.position = 'right')

dimplots_patient = DimPlot(combined, reduction = "proj.umap", group.by = "patient",cols = c(RColorBrewer::brewer.pal(n = 4, name = "Accent"))[4:1],pt.size=0.01,raster= F,order = F) + 
    ggtitle('All cells/nuclei ~ Patients') +
    guides(shape='none',color=guide_legend(override.aes = list(size=10), ncol=1)) +
    labs(color = "Patients") +
    umap_theme() + theme(plot.title = element_text(size = 42),legend.text = element_text(size=20),legend.title = element_text(size=28), legend.position = 'right')


png(file.path(params$datapath,'../results/Figures_March2024/UMAP_allcells.png'),width = 3000,height = 2250)
(dimplots_allcells | dimplots_type) / (dimplots_method | dimplots_patient) + plot_annotation(tag_levels = 'A') & theme(plot.tag = element_text(size = 34))
dev.off()
```

```{r per patient, message= T}
filenames_df = data.frame(filenames = filenames,
                          patient = sapply(strsplit(filenames, "-",fixed = T), "[",1),
                          method = sapply(strsplit(filenames, "_",fixed = T), "[",2),
                          tissue = sapply(strsplit(filenames, "-",fixed = T), "[",2))

filenames_df$method = gsub('all','Cell',filenames_df$method)
filenames_df$method = gsub('SN','Nucleus',filenames_df$method)
filenames_df$tissue = gsub('CST','Normal',filenames_df$tissue)
filenames_df$tissue = gsub('CTT','Tumor',filenames_df$tissue)
filenames_df$patient = gsub('4','patient1',filenames_df$patient)
filenames_df$patient = gsub('6','patient2',filenames_df$patient)
filenames_df$patient = gsub('8','patient3',filenames_df$patient)
filenames_df$patient = gsub('10','patient4',filenames_df$patient)

#cell types general plots 
barplots_general = list()
plot_UMAPS = list()
for(i in 1:24){
  data_summarised = seurat.objects[[i]]@meta.data %>% 
      group_by(predicted.ann_level_1,patient) %>%
      summarise(n = length(predicted.ann_level_1),.groups='keep')

  legend_position = ifelse(i ==9|i ==10,"right",'none')
  print(paste0('i ', legend_position)) 
  
barplots_general[[i]] = ggplot(data=data_summarised, aes(x='', y=n,fill=predicted.ann_level_1)) +
    geom_bar(stat="identity")+scale_fill_manual('Cell Types (level1)',values=c(RColorBrewer::brewer.pal(n = 5, name = "Set1")))+theme_classic()+theme(axis.title = element_text(size = 8),legend.position = legend_position) +ylab('Total Number of cells') + xlab('')

###plots
  plot_UMAP=DimPlot(seurat.objects[[i]], reduction = "proj.umap", group.by = "predicted.ann_level_1",cols = brewer.pal(n = 5, name = "Set1"),pt.size=0.2) + ggtitle(paste0(filenames_df$patient[i],' ~ ',filenames_df$method[i])) + guides(shape='none') +
    theme(legend.position = "none",axis.title = element_text(size = 8),axis.text =element_text(size = 8),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank())
 
  plot_UMAP$labels$colour = 'Cell types (level1)' 
  
  plot_UMAPS[[i]] = plot_UMAP
}
  
#save plots  
  pdf(file.path(params$datapath,'../results/Figures_March2024/FigureS3.pdf'),width = 9,height = 10.5) 
 (plot_UMAPS[[7]] + barplots_general[[7]] + plot_UMAPS[[9]] + barplots_general[[9]] + plot_layout(widths = c(10,1,10,1))) / 
 (plot_UMAPS[[13]] + barplots_general[[13]] + plot_UMAPS[[15]] + barplots_general[[15]] + plot_layout(widths = c(10,1,10,1))) / 
 (plot_UMAPS[[19]] + barplots_general[[19]] + plot_UMAPS[[21]] + barplots_general[[21]] + plot_layout(widths = c(10,1,10,1))) /
 (plot_UMAPS[[1]] + barplots_general[[1]] + plot_UMAPS[[3]] + barplots_general[[3]] + plot_layout(widths = c(10,1,10,1)))
  dev.off()

   pdf(file.path(params$datapath,'../results/Figures_March2024/FigureS5.pdf'),width = 9,height = 10.5) 
 (plot_UMAPS[[11]] + barplots_general[[11]] + plot_UMAPS[[10]] + barplots_general[[10]] + plot_layout(widths = c(10,1,10,1))) / 
 (plot_UMAPS[[16]] + barplots_general[[16]] + plot_UMAPS[[18]] + barplots_general[[18]] + plot_layout(widths = c(10,1,10,1))) / 
 (plot_UMAPS[[22]] + barplots_general[[22]] + plot_UMAPS[[24]] + barplots_general[[24]] + plot_layout(widths = c(10,1,10,1))) /
 (plot_UMAPS[[4]] + barplots_general[[4]] + plot_UMAPS[[6]] + barplots_general[[6]] + plot_layout(widths = c(10,1,10,1)))
  dev.off()


```


# feature plots
* Feature plots avec les marqueurs EPCAM, CLDN5, COL1A2 et PTPRC pourraient aussi illustrer les grands compartiments Epithelial, Endothelial, Stromal et Immune (panels E, F, G, H).
```{r feature plots, message= T}

combined_permethod = readRDS(file.path(params$datapath,'combined_permethod_pertypeN24.rds'))

for(i in 1:2) {
  combined_permethod[[i]] <- AddModuleScore(object = combined_permethod[[i]], features = list(Immune), name = "Immune_module",nbin = 12,ctrl = 100)
  combined_permethod[[i]] <- AddModuleScore(object = combined_permethod[[i]], features = list(Epithelial), name = "Epithelial_module",nbin = 12,ctrl = 100)
  combined_permethod[[i]] <- AddModuleScore(object = combined_permethod[[i]], features = list(Endothelial), name = "Endothelial_module",nbin = 12,ctrl = 100)
  combined_permethod[[i]] <- AddModuleScore(object = combined_permethod[[i]], features = list(Stroma), name = "Stroma_module",nbin = 12,ctrl = 100)
}

endo = 'CLDN5'
epi = 'EPCAM'
immune = 'PTPRC'
stroma = 'COL1A2'
cols = c('blue','gray','#FF0000')

fplots = list()

for(i in 1:2)
  {
  fplots[[(i*5) - 4]] = DimPlot(combined_permethod[[i]], reduction = "proj.umap", group.by = "predicted.ann_level_1",cols = col,pt.size=0.01,raster= F,order = F, label = T, label.size = 7, repel = T)  + ggtitle(ifelse(i==1,'Normal ~ Cell','Normal ~ Nucleus')) + umap_theme() + theme(legend.position = "none")

  fplots[[(i*5) - 3]] = FeaturePlot(combined_permethod[[i]], reduction = "proj.umap",features = 'Immune_module1',pt.size=0.01,raster= F,order = T,cols = cols) + theme(legend.position = "right") + ggtitle('Immune markers') + umap_theme()
  fplots[[(i*5) - 2]] = FeaturePlot(combined_permethod[[i]], reduction = "proj.umap",features = 'Epithelial_module1',pt.size=0.01,raster= F,order = T,cols = cols) + theme(legend.position = "right") + ggtitle('Epithelial markers')+ umap_theme()
  fplots[[(i*5) - 1]] = FeaturePlot(combined_permethod[[i]], reduction = "proj.umap",features = 'Endothelial_module1',pt.size=0.01,raster= F,order = T,cols = cols) + theme(legend.position = "right") + ggtitle('Endothelial markers')+ umap_theme()
  fplots[[(i*5)]] = FeaturePlot(combined_permethod[[i]], reduction = "proj.umap",features = 'Stroma_module1',pt.size=0.01,raster= F,order = T,cols = cols) + theme(legend.position = "right") + ggtitle('Stroma markers')+ umap_theme()
}

#save plot as png (because a .pdf as a vectorized image is too big)
png(file.path(params$datapath,'../results/Figures_March2024/Figure_S2.png'),width = 1560,height = 600)
((fplots[[1]]|fplots[[2]]|fplots[[3]]|fplots[[4]]|fplots[[5]]) /
(fplots[[6]]|fplots[[7]]|fplots[[8]]|fplots[[9]]|fplots[[10]])) + plot_annotation(tag_levels = 'A') + plot_layout(guides = 'collect')
dev.off()
```


# session info  
```{r session, message= T}
###session
sessionInfo()
```











